yapapi/executor/__init__.py

```diff
@@ -557,22 +566,66 @@ async def start_worker(agreement: rest.market.Agreement, node_info: NodeInfo) ->
                             await batch.prepare()
                             cc = CommandContainer()
                             batch.register(cc)
+
+                            import colors
+
+                            print(colors.yellow("[executor] Batch:"), cc.commands())
+
                             remote = await act.send(
                                 cc.commands(), stream_output, deadline=batch_deadline
                             )
                             cmds = cc.commands()
                             emit(events.ScriptSent(agr_id=agreement.id, task_id=task_id, cmds=cmds))
 
-                            async for evt_ctx in remote:
-                                evt = evt_ctx.event(agr_id=agreement.id, task_id=task_id, cmds=cmds)
-                                emit(evt)
-                                if isinstance(evt, events.CommandExecuted) and not evt.success:
-                                    raise CommandExecutionError(evt.command, evt.message)
+                            async def get_batch_results():
+
+                                print(colors.yellow("[get_batch_results] Waiting for results..."))
 
-                            emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
-                            await batch.post()
-                            emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
-                            await accept_payment_for_agreement(agreement.id, partial=True)
+                                results = []
+                                async for evt_ctx in remote:
+                                    evt = evt_ctx.event(
+                                        agr_id=agreement.id, task_id=task_id, cmds=cmds
+                                    )
+                                    emit(evt)
+                                    results.append(evt)
+                                    if isinstance(evt, events.CommandExecuted) and not evt.success:
+                                        raise CommandExecutionError(evt.command, evt.message)
+
+                                emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
+                                await batch.post()
+                                emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
+                                await accept_payment_for_agreement(agreement.id, partial=True)
+
+                                print(colors.yellow("[get_batch_results] Got results"))
+                                return results
+
+                            if batch.wait_for_results or batch.contains_init_step:
```

Comment:
why are we assuming that existence of the init step should make the call blocking?

Reply:
It's just to prevent the user from sending another exe script before `deploy` and `start` commands are executed.

Reply:
hmm.... I guess we might want to change the interface re: the init step anyway but for now, maybe we should add `assert not batch.contains_init_step` in the `else` clause here?